compiler/lib/src/main/scala/codegen/CppWriter/CppWriterState.scala

```diff
@@ -16,6 +16,8 @@ case class CppWriterState(
   pathPrefixes: List[String] = Nil,
   /** The default string size */
   defaultStringSize: Int = CppWriterState.defaultDefaultStringSize,
+  /** Whether or not to emit template code */
+  template: Boolean = false,
```

Comment:
Let's make this a variant type, not a Boolean. We should have one variant for each mode. I think there are two modes:
1. AUTOCODE: Generate F Prime autocode
2. IMPL_TEMPLATE: Generate a template for the hand-coded implementation

For the variant type, in Scala 3 we can use an enum.

Reply:
Actually if we have two separate visitors for the two modes, we may not need a mode variable, right? We could just call the appropriate visitor from the top-level code.

Reply:
Maybe we could make CppWriter an abstract class and have two subclasses, AutocodeCppWriter and ImplCppWriter.

Reply:
I think we should have the two separate visitors, but leave CppWriter as is (minus the visitor functions), since it has some functions that are used by the other cpp writer classes.

Reply:
> it has some functions that are used by the other cpp writer classes

OK, could we move those functions to CppWriterUtils?

Some helper functions are private and I think those should be made protected members of CppWriter so that AutocodeCppWriter and ImplCppWriter can inherit from CppWriter and use them.

Reply:
Ok, this sounds good. I also realized those functions can just stay in a companion object of CppWriter.